src/cargo/ops/cargo_add/mod.rs

```diff
@@ -364,24 +364,28 @@ fn resolve_dependency(
     };
     selected_dep = populate_dependency(selected_dep, arg);
 
-    let old_dep = get_existing_dependency(manifest, selected_dep.toml_key(), section)?;
-    let mut dependency = if let Some(mut old_dep) = old_dep.clone() {
-        if old_dep.name != selected_dep.name {
-            // Assuming most existing keys are not relevant when the package changes
-            if selected_dep.optional.is_none() {
-                selected_dep.optional = old_dep.optional;
+    let get_dependency = |mut selected_dep: Dependency| {
```

Comment:
> I created a fix for this issue. I did not follow the route of doing fuzzy lookups within cargo add, but just repeat the workspace existing dependency lookup and workspace dependency lookup when the latest dependency lookup returns a different name.

This approach is pretty complicated, requiring the code to duplicate its logic to cycle things back through.  At minimum, it seems like we could look in the registry from the beginning and then pass the literal name given and the registry normalized name around to do lookups.

Though that will also have problems.  This is using crates.io to normalize package names regardless of whether their source is crates.io or not.  We really should be normalizing as we go.  Unlike the registry, we don't need to try a bunch of combinations but we can extract the names from a table, normalize them to either `-` or `_`, normalize our value to `-` or `_` to match and see what the final result would be.

Reply:
Sorry, for the late reply.

> This approach is pretty complicated, requiring the code to duplicate its logic to cycle things back through.

I agree that this is not pretty. On the other hand it checks some boxes: It avoids duplicating logic in the code, It is pretty minimal, It avoids querying the registry if it is not needed, It applies the same fuzzy search logic as the registry.

> At minimum, it seems like we could look in the registry from the beginning and then pass the literal name given and the registry normalized name around to do lookups.

This has the problem that we always have to lookup the registry. This breaks a lot of tests by changing their outputs  by adding "Updating crates.io index" lines.

> This is using crates.io to normalize package names regardless of whether their source is crates.io or not.

Mhh, not sure whether I understand this properly. Sources of dependencies can be workspace dependencies, the members of the workspace and crates.io. As far as I understand only crates.io does fuzzy lookups. In which scenario crates.io be used to normalize package names that are not from crates.io?

> We really should be normalizing as we go.

I understand that you want to try several versions of the name - the original one, one that normalizes to underscores and one that normalizes to dashes, right? I can try that.

Reply:
> Mhh, not sure whether I understand this properly. Sources of dependencies can be workspace dependencies, the members of the workspace and crates.io. As far as I understand only crates.io does fuzzy lookups. In which scenario crates.io be used to normalize package names that are not from crates.io?

git dependencies

> I understand that you want to try several versions of the name - the original one, one that normalizes to underscores and one that normalizes to dashes, right? I can try that.

That should be fine as that is all we are checking for the registry today, see #12083

Reply:
I updated this PR to check permutations